### ISSUE TITLE
Decode carries a fixed ~60ms per-token cost: 135M runs at 14 tok/s against llama.cpp's 200, at every thread count

### ISSUE THREAD

**Issue by antonellof** [antonellof/ferrox#128]
Measured 2026-09-04 on a rented, idle 20-core Cortex-A725 (aarch64, ARMv9.2 with `i8mm`), ferrox and llama.cpp built and run on the same box, same GGUF, CPU only.

## The number

`SmolLM2-135M-Instruct-Q8_0`, `tg128`:

| threads | ferrox rayon | ferrox spin | llama.cpp |
|---|---|---|---|
| 4 | 13.22 | 15.83 | 189.86 |
| 8 | 14.49 | 13.71 | |
| 19 | 14.46 | 9.19 | 203.84 |

ferrox sits at **13 to 15 tok/s at every thread count**. llama.cpp does **190 to 204** on the same host and file. That is a **~14x** decode gap on a 135M model.

## Why this is not #27

[#27](https://github.com/antonellof/ferrox/issues/27) says CPU decode is scheduling-bound, and on 3B and 8B it clearly is: the persistent pool is +123% and +87% there, enough to overtake llama.cpp. None of that helps here.

- **Flat in thread count.** 4, 8 and 19 threads all give ~14 tok/s. Fork-join cost scales with worker count; this does not, so it is not fork-join.
- **Unmoved by the pool.** Switching to the persistent pool changes it by a few percent at 4 threads and makes it *worse* at 19.
- **Flat in model size, in absolute terms.** Per token: 135M ~69 ms, 3B ~96 ms, 8B ~150 ms against llama.cpp's ~5 ms, ~56 ms, ~110 ms. The difference is roughly **40 to 65 ms per token regardless of size**, which is the signature of a fixed cost paid once per token rather than work proportional to the weights.

At 8B that constant is 27% of the token. At 135M it is 93% of it, which is why the ratio explodes as the model shrinks.

## What to look for

Something done once per decoded token whose cost does not depend on the weights: a per-token allocation or zeroing of a scratch buffer sized by vocabulary or context, a KV-cache bookkeeping pass, a per-token `Vec` build in the sampler, logits copied more than once, or a per-token probe/lookup that is not hoisted. `MIN_TASK_MACS` and the pool are both about *parallel* cost and cannot see this.

A profile of one decode step at 135M with 1 thread would name it immediately; the constant is 90%+ of the step at that size, so it cannot hide.

## Why it matters

Small models are exactly where a user notices tok/s, and 135M to 1B is the size range for on-device and draft-model use. **Speculative decoding runs a small model as the drafter**, so a fixed per-token cost is paid on every draft token too, which caps what `-d` can ever win.

The M2 Pro ledger does not show this: at 6 threads and 48 tok/s the constant is a smaller share, so the row looks merely slow rather than structurally wrong.

**Comment by antonellof** [comment]
## Profiled. It is rayon handoff latency, and it is 82–87% of the token

You asked for a profile of one decode step at 135M with one thread. Here it is, and it reproduces on a **second machine and a different microarchitecture** (Apple M2 Pro, macOS), so it is not specific to the rented Cortex-A725.

`SmolLM2-135M-Instruct-Q8_0`, CPU only, `sample` over the main thread, leaf self-time:

| | `-t 1` | `-t 6` |
|---|---:|---:|
| **`__psynch_cvwait`** | **87.4%** | **82.4%** |
| `causal_gqa_attention_softmax` | 10.5% | 14.3% |
| rayon `for_each` consumer | 0.8% | 1.4% |
| everything else | ~1.3% | ~1.9% |

The caller chain is `rayon_core::latch::LockLatch::wait_and_reset` → `_pthread_cond_wait` → `__psynch_cvwait`. **The main thread is parked on a condition variable waiting for rayon for most of every token**, and it is parked at one thread, where there is nothing to wait for.

### The arithmetic

At `-t 1` ferrox runs 36.86 tok/s = 27.1 ms/token, of which 87.4% is waiting = **23.7 ms**. llama.cpp on the same box and file runs 247.99 tok/s = **4.03 ms for the entire token**. The waiting alone is 5.9× llama.cpp's whole step.

That is the fixed cost this issue describes, and every symptom follows from it: a condvar handoff is a synchronization round trip, so it does not shrink with thread count, does not scale with weights, and is paid once per parallel region per token.

### Neither pool fixes it

| pool | `-t 1` | `-t 6` |
|---|---:|---:|
| rayon | 34.98 | **54.65** |
| spin | **38.93** | 33.77 |
| llama.cpp | 247.99 | 300.05 |

`FERROX_CPU_POOL=spin` helps 11% at one thread and is **38% worse** at six — which is what this issue already reported at 19 threads, now reproduced at 6. Spin changes how the workers wait; it does not remove the handoff, and at realistic thread counts the spinning costs more than the condvar did.

### Why this is structural, not tuning

ferrox dispatches a parallel region **per matmul**: the main thread forks, blocks, and joins, roughly 7 times per layer. llama.cpp does not do this. Its threads all run the graph and meet at a barrier per node — there is no main thread handing work to workers and sleeping.

So the fix is not a better pool or a better `MIN_TASK_MACS`. It is removing the per-matmul fork-join from the decode path: either run the whole decode step inside one parallel region with barriers between stages, or run small matmuls inline on the calling thread and never dispatch at all.

`MIN_TASK_MACS` is `1 << 16` = 65536, and SmolLM2's `q_proj` is 576 × 576 = 331k MACs, so every projection clears the bar and dispatches. At ~30 layers that is roughly 200 dispatches per token, which at even 100 µs of handoff each accounts for the whole constant.

### Reproducing

```
FERROX_METAL=0 ./target/release/ferrox bench \
  -m models/hf_test/SmolLM2-135M-Instruct-Q8_0.gguf -p 0 -n 512 -r 4 -t 1 --max-load 0 &
sample $(pgrep -n -f "release/ferrox bench") 8 -mayDie
```

The `--max-load 0` is deliberate: this is a profile shape, not a publishable tok/s.

**Comment by antonellof** [comment]
## Correcting my own comment above: fork-join is 6.7% of the token, not 82%

My previous comment read "82–87% of the main thread is `__psynch_cvwait`" and concluded the fixed cost is rayon handoff. **The share is real; the conclusion from it was wrong**, and it is the same mistake this repo already wrote into `benchmarks/RESULTS.md` as a trap after making it on Metal: *a share of a profile is not a marginal cost*. The main thread being parked mostly means it is waiting for workers doing real work, which is what a fork-join design looks like when it is working.

### Measured directly

Injecting N extra `par::indices` regions per `WeightMatrix::apply` and taking the slope, interleaved, SmolLM2-135M Q8_0, `-t 6`:

| | tok/s | ms/token |
|---|---:|---:|
| N=0 | 58.03 | 17.23 |
| N=8 | 37.82 | 26.44 |

1680 extra regions cost 9.21 ms → **5.48 µs per fork-join region**.

At ~210 regions per token (30 layers × ~7 matvecs), the entire dispatch cost is **1.15 ms, or 6.7%** of the token. **Removing every fork-join in the model would leave 16.08 ms against llama.cpp's 3.33 ms — still 4.8× behind.**

So scheduling cannot be what this issue is about, and neither `MIN_TASK_MACS`, the persistent pool, nor an inline fast path can reach it. I built the inline path (skip rayon when `with_min_len` has already collapsed the work to one chunk), measured it, got nothing, and dropped it.

### What the number actually says

Convert to the resource. SmolLM2-135M Q8_0 is 138 MB read per token:

| | ms/token | achieved | share of ~100 GB/s |
|---|---:|---:|---:|
| ferrox | 17.23 | 8.0 GB/s | **8.0%** |
| llama.cpp | 3.33 | 41.5 GB/s | **41.5%** |

Decode is memory-bound, and ferrox is at **8% of the machine's bandwidth where llama.cpp is at 41%**. That is a kernel efficiency problem — the same shape as the CUDA decode finding in #133, not a scheduling one.

### What that means for this issue

The "fixed cost" framing came from the constant looking flat across model sizes. A per-token cost that is 5× too slow at every size *also* looks flat in absolute terms when the models are small. The thing to chase is the dot-product kernel's achieved bandwidth at batch=1, not the thread pool.

Worth checking first, before any kernel work: whether the 135M case is running the repacked interleaved tier at all, or falling back. That tier is what carries the aarch64 numbers (see #152), and a small model with an unusual shape is exactly where a repack predicate can quietly decline.

**Comment by antonellof** [comment]
## Found it: `FERROX_CPU_INT_DOT` costs Q8_0 decode 2× to 5×, and it is on by default here

Following the lead at the end of my last comment — "check whether the repacked tier is engaged" — the answer is that it is engaged, and for **Q8_0 decode it is the problem, not the fix**.

Decode tok/s on an M2 Pro, `-t 6`, CPU only, with the flag on vs off:

| kind | model | on | off | |
|---|---|---:|---:|---|
| **Q8_0** | SmolLM2-135M | 59.3 | **119.9** | **off 2.0×** |
| **Q8_0** | Qwen2.5-0.5B | 19.2 | **73.1** | **off 3.8×** |
| **Q8_0** | TinyLlama-1.1B | 8.2 | **41.3** | **off 5.0×** |
| Q4_0 | OLMoE-1B-7B | **50.2** | 26.6 | on 1.9× |
| Q4_K | Llama-3.2-1B | **84.6** | 21.3 | on 4.0× |
| Q4_K | Llama-3.2-3B | **36.7** | 8.3 | on 4.4× |
| Q5_K | Llama-3.2-1B | **70.6** | 23.5 | on 3.0× |
| Q6_K | Llama-3.2-1B | **61.9** | 25.6 | on 2.4× |
| IQ4_XS | Llama-3.2-1B | **37.9** | 32.4 | on 1.2× |

**Q8_0 is the only kind that loses, and it is the kind this issue is filed about.** `int_dot_is_a_win_here()` returns `cfg!(target_arch = "aarch64")`, so on the aarch64 host this issue was measured on, the flag is ON — and SmolLM2-135M is Q8_0. The ledger's worst aarch64 CPU row is `SmolLM2-135M Q8_0` at 8.2×.

Prefill is the opposite and wants the path for every kind measured (Q8_0 4.6×, Q4_K 15×), so this is one global boolean deciding three independent questions: prefill vs decode, quant kind, and architecture.

## Where it is NOT

I tried the obvious fix twice and both measured nothing, so recording it so nobody repeats it:

1. Guarding the `QuantKind::Q8_0` arm of the int-dot match in `WeightMatrix::apply` — SmolLM2 58 → 53 (worse).
2. Gating the whole `if cpu_int_dot_enabled()` block on the kind — SmolLM2 58 → 53, TinyLlama 8.2 → 8.1, i.e. the 5× is completely untouched.

So the cost is **not in the matvec dispatch**.

## Where it probably is

The penalty scales with model size, which a per-token cost would not do:

| model | weights | penalty |
|---|---:|---:|
| SmolLM2-135M | 0.14 GB | 2.0× |
| Qwen2.5-0.5B | 0.68 GB | 3.8× |
| TinyLlama-1.1B | 1.12 GB | 5.0× |

And the flag costs memory: RSS with it on is **1734 MB against a 1120 MB model file**, roughly 600 MB extra, which is the shape of the lazy `block_q8_0x4` repack holding a second copy of the weights. With the flag off the same benchmark had already finished before a 12-second sample could catch the process.

So the next step is `get_or_repack_q8x4`: whether the repack is cached as intended or being redone, and what it costs to build and then stream a second copy of the weights through a cache that was already holding the first.

That is a specific, testable place, and it is not the thread pool — which my earlier comment on this issue got wrong, and which the correction above documents.

**Comment by antonellof** [comment]
## Root cause: the FFN matrices are repacked on EVERY matvec

Instrumented `get_or_repack` and ran TinyLlama-1.1B Q8_0 decode for 32 tokens:

```
PROBE uncached repack #1     rows=5632 cols=2048
PROBE uncached repack #500   rows=5632 cols=2048
...
PROBE uncached repack #2500  rows=5632 cols=2048
PROBE cache hits #6000
```

`get_or_repack` takes `id: Option<MapId>` and on `None` does `return Arc::from(repack())` — **no cache lookup, no insert**. `WeightBytes::map_id()` returns `None` for `Owned` and `Shared` bytes. The attention projections are `Mapped` and hit the cache (6000 hits); the FFN gate/up/down (5632×2048) are not, so every one of ~2500 matvecs re-runs `pack_q8_0_matrix_x4` over an 11.5 MB matrix and allocates a fresh `Arc<[u8]>` for it.

That is the 5×, and it is why the penalty scales with model size — it is a per-matvec repack of the largest matrices in the model, not a per-token constant. It also explains the ~600 MB extra RSS.

Two questions for the fix, in order:
1. Why are those FFN tensors `Owned`/`Shared` rather than `Mapped` in the first place? A dense GGUF tensor should be a zero-copy mmap view. If the loader copies them, that is the bug to fix and the cache question disappears.
2. If some bytes legitimately cannot be `Mapped`, `get_or_repack` still needs a key for them — a `None` id must not mean "repack forever".

Guard for the fix: a test that a `WeightMatrix` over `Owned` bytes repacks **once** across N `apply` calls, which the current code fails on.
